runtime/src/accounts_db.rs

```diff
@@ -6990,6 +7003,62 @@ impl AccountsDb {
                 })
                 .sum();
 
+            // subtract data.len() from accounts_data_len for all old accounts that are in the index twice
+            if pass == 0 {
+                let pubkeys_to_accounts_data_len = |pubkeys: &[Pubkey]| {
+                    let mut accounts_data_len_from_duplicates = 0;
+                    pubkeys.into_iter().for_each(|pubkey| {
+                        if let Some(entry) = self.accounts_index.get_account_read_entry(pubkey) {
+                            let slot_list = entry.slot_list();
+                            if slot_list.len() < 2 {
+                                return;
+                            }
+                            let mut slot_list = slot_list.clone();
+                            slot_list.sort_unstable_by(|a, b| a.0.cmp(&b.0));
+                            assert!(slot_list[0].0 < slot_list[1].0);
+                            slot_list
+                                .into_iter()
+                                .rev()
+                                .skip(1)
+                                .for_each(|(slot, account_info)| {
+                                    let maybe_storage_entry = self
+                                        .storage
+                                        .get_account_storage_entry(slot, account_info.store_id);
+                                    let mut accessor = LoadedAccountAccessor::Stored(
+                                        maybe_storage_entry
+                                            .map(|entry| (entry, account_info.offset)),
+                                    );
+                                    let loaded_account = accessor.check_and_get_loaded_account();
+                                    let account = loaded_account.take_account();
+                                    accounts_data_len_from_duplicates += account.data().len();
+                                });
+                        }
+                    });
+                    accounts_data_len_from_duplicates as u64
+                };
+
+                let mut timer = Measure::start("handle accounts data len duplicates");
+                let mut unique_pubkeys = HashSet::<Pubkey>::default();
+                self.uncleaned_pubkeys.iter().for_each(|entry| {
+                    entry.value().iter().for_each(|pubkey| {
+                        unique_pubkeys.insert(*pubkey);
+                    })
+                });
+                let accounts_data_len_from_duplicates = unique_pubkeys
+                    .into_iter()
+                    .collect::<Vec<_>>()
+                    .par_chunks(4096)
+                    .map(pubkeys_to_accounts_data_len)
+                    .sum();
+                accounts_data_len.fetch_sub(accounts_data_len_from_duplicates, Ordering::Relaxed);
```

Comment:
So now this is the main logic, of a `map()` and a `sum()`, which rayon handles. Then after that number is totaled up, only need the single/final atomic subtraction.

I tested this impl out by printing the accounts data len from generate_index(), and the immediately calling the `Bank::get_total_accounts_stats()` afterwards, which uses `scan_accounts()` under the hood, and asserted the values were always equal.